File: contests/lc/1912.DesignMovieRentalSystem/design-movie-rental-system.py
import collections
import heapq
from typing import Dict, List, Optional, Tuple


class MovieRentingSystem:
    def __init__(self, n: int, entries: List[List[int]]):
        self.unrented: Dict[int, List[Tuple[int, int, int]]] = collections.defaultdict(list)
        self.price_lookup: Dict[Tuple[int, int], int] = {}
        self.unrented_id: Dict[Tuple[int, int], Optional[int]] = {}
        self.rented_heap: List[Tuple[int, int, int, int]] = []
        self.rented_id: Dict[Tuple[int, int], Optional[int]] = {}
        self._next_entry_id = 0

        for shop, movie, price in entries:
            entry_id = self._allocate_entry_id()
            heapq.heappush(self.unrented[movie], (price, shop, entry_id))
            self.unrented_id[(shop, movie)] = entry_id
            self.rented_id[(shop, movie)] = None
            self.price_lookup[(shop, movie)] = price

    def search(self, movie: int) -> List[int]:
        result: List[int] = []
        temp: List[Tuple[int, int, int]] = []
        heap = self.unrented[movie]

        while heap and len(result) < 5:
            price, shop, entry_id = heapq.heappop(heap)
            if self.unrented_id.get((shop, movie)) != entry_id:
                continue
            result.append(shop)
            temp.append((price, shop, entry_id))

        for item in temp:
            heapq.heappush(heap, item)

        return result

    def rent(self, shop: int, movie: int) -> None:
        price = self.price_lookup[(shop, movie)]
        self.unrented_id[(shop, movie)] = None
        rented_entry_id = self._allocate_entry_id()
        self.rented_id[(shop, movie)] = rented_entry_id
        heapq.heappush(self.rented_heap, (price, shop, movie, rented_entry_id))

    def drop(self, shop: int, movie: int) -> None:
        price = self.price_lookup[(shop, movie)]
        self.rented_id[(shop, movie)] = None
        unrented_entry_id = self._allocate_entry_id()
        self.unrented_id[(shop, movie)] = unrented_entry_id
        heapq.heappush(self.unrented[movie], (price, shop, unrented_entry_id))

    def report(self) -> List[List[int]]:
        result: List[List[int]] = []
        temp: List[Tuple[int, int, int, int]] = []

        while self.rented_heap and len(result) < 5:
            price, shop, movie, entry_id = heapq.heappop(self.rented_heap)
            if self.rented_id.get((shop, movie)) != entry_id:
                continue
            result.append([shop, movie])
            temp.append((price, shop, movie, entry_id))

        for item in temp:
            heapq.heappush(self.rented_heap, item)

        return result

    def _allocate_entry_id(self) -> int:
        entry_id = self._next_entry_id
        self._next_entry_id += 1
        return entry_id


def main():
    # Example usage:
    entries = [[0, 1, 5], [0, 2, 6], [0, 3, 7], [1, 1, 4], [1, 2, 7], [2, 1, 5]]
    movieRentingSystem = MovieRentingSystem(3, entries)
    print(movieRentingSystem.search(1))  # return [1,0,2]. Shops 1,0 and 2 have an unrented copy of movie 1. Their prices are [4,5,5] respectively. The cheapest shops are [1,0] (price 4 and 5 respectively). Shop 2 is also included since there are less than 5 shops.
    movieRentingSystem.rent(0, 1)         # Rent movie 1 from shop 0. Unrented movies at shop 0 are now [2,3].
    print(movieRentingSystem.report())    # return [[0,1]]. The only rented movie is [1] from shop [0].
    movieRentingSystem.drop(0, 1)         # Drop off movie 1 at shop 0. Unrented movies at shop 0 are now [1,2,3].
    print(movieRentingSystem.search(1))   # return [1,0,2]. Shops 0,1 and 2 have an unrented copy of movie 1. Their prices are [5,4,5] respectively. The cheapest shops are [1,0] (price 4 and 5 respectively). Shop 2 is also included since there are less than 5 shops.
    movieRentingSystem.rent(1, 2)         # Rent movie 2 from shop 1. Unrented movies at shop 1 are now [1].
    movieRentingSystem.rent(2, 1)         # Rent movie 1 from shop 2. Unrented movies at shop 2 are now [].
    print(movieRentingSystem.report())    # return [[2,1],[1,2]]. The rented movies are [2] from shop [1] (price=7) and [1] from shop [2] (price=5). The cheapest is [1] from shop [2] followed by [2] from shop [1].


if __name__ == "__main__":
    main()

# Your MovieRentingSystem object will be instantiated and called as such:
# obj = MovieRentingSystem(n, entries)
# param_1 = obj.search(movie)
# obj.rent(shop,movie)
# obj.drop(shop,movie)
# param_4 = obj.report()

File: contests/lc/1912.DesignMovieRentalSystem/design-movie-rental-system.go
package designmovierentalsystem

import (
	"container/heap"
	"fmt"
)

type IndexKey struct {
	Shop  int
	Movie int
}

func (k IndexKey) String() string {
	return fmt.Sprintf("%d_%d", k.Shop, k.Movie)
}

type MovieRentingSystem struct {
	rentedMovieHeap   *Movies
	unrentedMovies    map[int]*Movies
	rentedMovies      map[string]*RentingMovie
	indexingKeyOffset map[string]*RentingMovie
}

func Constructor(n int, entries [][]int) MovieRentingSystem {
	unrentedMovies := make(map[int]*Movies)
	indexingMovies := make(map[string]*RentingMovie)

	for _, entry := range entries {
		if _, ok := unrentedMovies[entry[1]]; !ok {
			unrentedMovies[entry[1]] = &Movies{}
		}
		m := &RentingMovie{
			Shop:  entry[0],
			Movie: entry[1],
			Price: entry[2],
		}
		heap.Push(unrentedMovies[entry[1]], m)
		indexingMovies[IndexKey{
			Shop:  entry[0],
			Movie: entry[1],
		}.String()] = m
	}

	return MovieRentingSystem{
		indexingKeyOffset: indexingMovies,
		unrentedMovies:    unrentedMovies,
		rentedMovieHeap:   &Movies{},
		rentedMovies:      make(map[string]*RentingMovie),
	}

}

func (mrs *MovieRentingSystem) Search(movie int) []int {
	movies, ok := mrs.unrentedMovies[movie]
	if !ok {
		return nil
	}
	result := make([]int, 0, 5)
	eligibleMovies := make([]*RentingMovie, 0, 5)
	for movies.Len() > 0 && len(eligibleMovies) < 5 {
		item := heap.Pop(movies)
		movie := item.(*RentingMovie)
		eligibleMovies = append(eligibleMovies, movie)
		result = append(result, movie.Shop)

	}
	if len(eligibleMovies) == 0 {
		return nil
	}
	for _, e := range eligibleMovies {
		heap.Push(movies, e)
	}
	return result

}

func (mrs *MovieRentingSystem) Rent(shop int, movie int) {

	key := IndexKey{
		Shop:  shop,
		Movie: movie,
	}

	rentingMovie := mrs.indexingKeyOffset[key.String()]
	if rentingMovie == nil {
		return
	}

	heap.Remove(mrs.unrentedMovies[movie], rentingMovie.Offset)
	//remove from heap min
	// add into rentedMovie List
	rentedMovie := &RentingMovie{
		Shop:     shop,
		Movie:    movie,
		Price:    rentingMovie.Price,
		IsRented: true,
	}
	mrs.rentedMovies[key.String()] = rentedMovie
	heap.Push(mrs.rentedMovieHeap, rentedMovie)
}

func (mrs *MovieRentingSystem) Drop(shop int, movie int) {

	key := IndexKey{
		Shop:  shop,
		Movie: movie,
	}
	m, ok := mrs.rentedMovies[key.String()]
	if !ok {
		return
	}
	rentingMovie := &RentingMovie{
		Shop:  shop,
		Movie: movie,
		Price: m.Price,
	}
	mrs.indexingKeyOffset[key.String()] = rentingMovie

	heap.Push(mrs.unrentedMovies[movie], rentingMovie)

	heap.Remove(mrs.rentedMovieHeap, m.Offset)

	delete(mrs.rentedMovies, key.String())
}

func (mrs *MovieRentingSystem) Report() [][]int {
	result := make([][]int, 0, 5)
	historicals := make([]*RentingMovie, 0, 5)
	for mrs.rentedMovieHeap.Len() > 0 && len(historicals) < 5 {
		item := heap.Pop(mrs.rentedMovieHeap)
		m := item.(*RentingMovie)
		historicals = append(historicals, m)
		result = append(result, []int{m.Shop, m.Movie})
	}
	if len(historicals) == 0 {
		return nil
	}
	for _, item := range historicals {
		heap.Push(mrs.rentedMovieHeap, item)
	}
	return result
}

type RentingMovie struct {
	Shop     int
	Movie    int
	Price    int
	Offset   int
	IsRented bool
}

type Movies []*RentingMovie

// Len implements heap.Interface.
func (m Movies) Len() int {
	return len(m)
}

// Less implements heap.Interface.
func (m Movies) Less(i int, j int) bool {
	if !m[i].IsRented {
		if m[i].Price == m[j].Price {
			return m[i].Shop < m[j].Shop
		}
	} else if m[i].Price == m[j].Price {
		if m[i].Shop == m[j].Shop {
			return m[i].Movie < m[j].Movie
		}
		return m[i].Shop < m[j].Shop
	}

	return m[i].Price < m[j].Price
}

// Pop implements heap.Interface.
func (m *Movies) Pop() any {
	item := (*m)[m.Len()-1]
	(*m) = (*m)[:m.Len()-1]
	return item
}

// Push implements heap.Interface.
func (m *Movies) Push(x any) {
	item := x.(*RentingMovie)
	item.Offset = m.Len()
	(*m) = append((*m), item)
}

// Swap implements heap.Interface.
func (m Movies) Swap(i int, j int) {
	m[i].Offset, m[j].Offset = j, i
	m[i], m[j] = m[j], m[i]
}
